### DISCUSSION TITLE
Debug with serial output

### DISCUSSION THREAD

**Discussion by clarkbremer** [SpenceKonde/megaTinyCore#704] [Library discussion]
Standard practice with Arduino development is to use `println()` statements to debug your code.  But with the UPDI programmer (CH340 w/Schottky diode) plugged into one USB port, I've found that I need a second USB->serial cable (I had an FTDI laying around) to monitor the serial output.  And the IDE wants the programming port to be the same as the serial monitor port, which it's not.  My workaround was to tell the IDE the port of the UPDI programmer, and use a different serial monitor program that's not part of the IDE to monitor the debug output on a different port.  

Is this "just the way it is", or am I missing something stupid?

**Comment by hmeijdam** [comment]
I am doing exactly as you describe and have only the ground and RX wire of the FTDI connected. Gives me the additional benefit of using a [better serial monitor program](http://der-hammer.info/pages/terminal.html) than the one built into the Arduino IDE

**Comment by mosqu-ito** [comment]
Keep in mind that standard Arduino development involves a bootloader.  It uses a serial port which, when the bootloader is done with it, is available to your program for println or anything else.  Our great benefactor also gives us a serial bootloader.  If you use it, then the situation is truly analogous.  UPDI though, is dedicated programming/debugging hardware with its own protocols and commands, and "debugging" refers to a system far more sophisticated than mere println statements in your code, this is real hardware debugging.  As much as it would be sweet to free up an extra haplex UART, it seems unlikely that Atmel offers us a mechanism with which to wrench out the one inside the UPDI hardware.  I would be as happy as anyone to be wrong about that.
